File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root: return True
        return self.helper(root, float('-inf'), float('inf'))
    def helper(self, node, leftBound, rightBound):
        if not node: return True
        if node.val <= leftBound or node.val >= rightBound:
            return False
        return self.helper(node.left, leftBound, node.val) and self.helper(node.right, node.val, rightBound)

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return helper(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
    private boolean helper(TreeNode node, long  leftBound, long rightBound){
        if (node == null) return true;
        if ((node.val <= leftBound) || (node.val >= rightBound)) return false;
        
        return helper(node.left, leftBound, node.val) && helper(node.right, node.val, rightBound);
    }
}